src/main/java/fr/baretto/ollamassist/chat/ui/AskToChatAction.java

```diff
@@ -1,46 +1,117 @@
 package fr.baretto.ollamassist.chat.ui;
 
+import dev.langchain4j.data.message.AiMessage;
 import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.model.output.FinishReason;
+import dev.langchain4j.service.TokenStream;
 import fr.baretto.ollamassist.chat.service.OllamaService;
+import lombok.Builder;
 import lombok.extern.slf4j.Slf4j;
 
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
 
 @Slf4j
 public class AskToChatAction implements ActionListener {
     private final PromptPanel promptPanel;
     private final MessagesPanel outputPanel;
     private final Context context;
+    private ChatThread currentChatThread;
 
     public AskToChatAction(PromptPanel promptInput, MessagesPanel outputPanel, Context context) {
         this.promptPanel = promptInput;
         this.outputPanel = outputPanel;
         this.context = context;
     }
 
-    private static void logException(Throwable throwable) {
+    private void logException(Throwable throwable) {
         log.error("Exception: " + throwable);
+        done(ChatResponse.builder().finishReason(FinishReason.OTHER).aiMessage(AiMessage.from(throwable.getMessage())).build());
     }
 
     @Override
     public void actionPerformed(ActionEvent e) {
         String userMessage = promptPanel.getUserPrompt();
+        if (userMessage.isEmpty()) {
+            return;
+        }
+
+        if (currentChatThread != null) {
+            currentChatThread.stop();
+        }
+        outputPanel.cancelMessage();
         outputPanel.addUserMessage(userMessage);
         outputPanel.addNewAIMessage();
 
-        new Thread(() -> context.project()
-                .getService(OllamaService.class)
-                .getAssistant()
-                .chat(userMessage)
+        currentChatThread = ChatThread.builder()
+                .tokenStream(context.project()
+                        .getService(OllamaService.class)
+                        .getAssistant()
+                        .chat(userMessage))
                 .onNext(this::publish)
-                .onError(AskToChatAction::logException)
+                .onError(this::logException)
                 .onCompleteResponse(this::done)
-                .start()
-        ).start();
+                .build()
+                .start();
         promptPanel.clear();
     }
 
+    @Builder
+    private static class ChatThread {
```

Comment:
Good idea! I was hesitating on how to implement the feature. Too bad that Langchain4j doesn’t allow stopping token generation.